### ISSUE TITLE
Fake setTimeout should "fail" on too large delays

### ISSUE THREAD

**Issue by fiddur** [sinonjs/lolex#67]
Both browsers and Node.js handles timers with int32 and overflows on values above 2147483647.  It would be good if lolex in some way fails in that case too.

Background:

I [read about](https://nodejs.org/api/timers.html#timers_settimeout_callback_delay_arg) the maximum delay on `setTimeout`, and prepared a `farTimeout` to just recursively wrap the  `setTimeout` with max timeout multiple times.  But, using test first approach, I noticed that the `setTimeout` handled 90 days without problem, even though the docs said max ~25 days.  Digging deeper, I realized that the sinon.useFakeTimers gave me a `setTimeout` that handled larger numbers than Node.js `setTimeout`.

Suggested solution:

I think something like this in `addTimer` function would make it behave more like the actual `setTimeout`:

``` javascript
const maxTimeout = 2147483647
if (timer.delay && timer.delay > maxTimeout) {
    timer.delay = (timer.delay + maxTimeout) % (maxTimeout * 2) - maxTimeout
}
```

…not making a pull request right now though since this was quite quickly thrown together; could make one if you think this is a good solution.

**Comment by fatso83** [comment]
Seems reasonable. But the if is unneeded, so the math could be simplified to

```javascript
// overflow if timer > maxTimeout
timer.delay = timer.delay % maxTimeout;
}
```

**Comment by fiddur** [comment]
I placed it in `addTimer` rather than `setTimeout` since there are similar limitations to other timers, but in `addTimer` there isn't always a `delay`?

The `+ max … - max` is to mimic the int32 overflow, that will sometimes render a negative value (making the timer trigger immediately).

**Comment by fatso83** [comment]
There are "timers" like `setImmediate` and `nextTick` that does not take arguments, so that might be why.

**Comment by fatso83** [comment]
Was there a PR coming?

**Comment by fiddur** [comment]
I'm sorry, my vacation got in the way!  I'll get to it this week I hope!

**Comment by fatso83** [comment]
New summer, new hopes :-)

**Comment by benjamingr** [comment]
There are a few edge cases: 

Node does:
> When delay is larger than 2147483647 or less than 1, the delay will be set to 1.

Timers API say:
> If timeout is less than 0, then set timeout to 0.

We don't handle this case at the moment - can PR to handle. We do handle it not being passed.

The spec itself does _not_ require that the timeout is under 2**32 nor does it require that the timers are overflown in a particular way.

MDN claims:
> Browsers including Internet Explorer, Chrome, Safari, and Firefox store the delay as a 32-bit signed integer internally. This causes an integer overflow when using delays larger than 2147483647 (about 24.8 days), resulting in the timeout being executed immediately.

We can definitely do that. 

> Timers can be nested; after five such nested timers, however, the interval is forced to be at least four milliseconds.

We can likely handle this case, but it's not very interesting (since order is still guaranteed)

**Comment by benjamingr** [comment]
Turns out this is actually well defined in WebIDL https://heycam.github.io/webidl/#abstract-opdef-converttoint (thanks to @annavk for pointing me in the right direction!).


> 9. Set x to ! IntegerPart(x).
> 10. Set x to x modulo 2bitLength.
> 11. If signedness is "signed" and x ≥ 2bitLength − 1, then return x − 2bitLength.
> 12. Otherwise, return x.

**Comment by benjamingr** [comment]
cc @elad-nach (have you started working on it by the way?)

**Comment by elad-nach** [comment]
this is implemented as part of the V8, why do you want to write a javascript implementation to something which is already implemented in the underlying engine?

**Comment by benjamingr** [comment]
@elad-nach what do you mean? lolex runs in many engines (unless you're suggesting to lift some part from V8), and even in V8 it's just how numbers work - and not implemented - it was specced to behave consistently with two's complement number arithmetic. 

Your current code sets the timer to 1ms no matter what - if you look at @fatso83's implementation it actually follows the spec I linked to above in https://github.com/sinonjs/lolex/issues/67#issuecomment-224914479 - but I don't think Chrome or any of the browsers do (overflow) - I think they all actually run the timer immediately.

An easy test for this would be to run with `2 ** 31 - 1 + 1000` and see that it runs immediately and not after 1000ms. @fatso83 - I'm leaning towards liking @elad-nach's current implementation better than the spec compliant one - what do you say?

**Comment by elad-nach** [comment]
NodeJS spec specifies that above delays above 2 ** 31 are set to 1. Also ones below 1.

**Comment by benjamingr** [comment]
@elad-nach yes, but the DOM spec suggests otherwise so we _might_ want to if/else and set timers accordingly - although this is already an edge case and I think setting them to 1 is a reasonable default for web too.

(I would throw on timers that are too long, but meh)

**Comment by fatso83** [comment]
Closed by #124
